### PR TITLE
feat: Allow Traffic shaping through header based routing for ALB

### PR DESCRIPTION
**Pull request by perenesenko**
Signed-off-by: Andrii Perenesenko <andrii.perenesenko@gmail.com>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

File: utils/ingress/ingress.go [hunk 1]
```diff
@@ -142,6 +162,26 @@ func hasLegacyIngressRuleWithService(ingress *extensionsv1beta1.Ingress, svc str
...
+type ManagedALBAnnotations map[string]ManagedALBAnnotation
+
+type ManagedALBAnnotation []string
+
+// String outputs a string of all the managed ALB annotations that is stored in the Ingress's annotations
+func (m ManagedALBAnnotations) String() string {
+	return string(json.MustMarshal(m))
+}
+
+func NewManagedALBAnnotations(json string) (ManagedALBAnnotations, error) {
```
Comment (zachaller) [inline utils/ingress/ingress.go:174]: Is this function even used, I only see it in tests?

File: utils/ingress/wrapper.go [hunk 2]
```diff
@@ -149,6 +173,85 @@ func (i *Ingress) SetAnnotations(annotations map[string]string) {
...
+					Name: actionName,
+					Port: v1.ServiceBackendPort{
+						Name: "use-annotation",
+					},
+				},
+			},
+		}
+		for _, rule := range i.ingress.Spec.Rules {
+			rule.HTTP.Paths = append(rule.HTTP.Paths[:1], rule.HTTP.Paths[0:]...)
+			rule.HTTP.Paths[0] = p
```
Comment (zachaller) [inline utils/ingress/wrapper.go:199]: This should be in the order of managedRoutes pretty sure if this is the route for headers we should place it on top of any existing path but ordered by managedRoutes order.

File: utils/ingress/wrapper.go [hunk 3]
```diff
@@ -149,6 +173,85 @@ func (i *Ingress) SetAnnotations(annotations map[string]string) {
...
+			Path:     "/*",
+			PathType: &t,
+			Backend: v1beta1.IngressBackend{
+				ServiceName: actionName,
+				ServicePort: intstr.FromString("use-annotation"),
+			},
+		}
+		for _, rule := range i.legacyIngress.Spec.Rules {
+			rule.HTTP.Paths = append(rule.HTTP.Paths[:1], rule.HTTP.Paths[0:]...)
+			rule.HTTP.Paths[0] = p
```
Comment (zachaller) [inline utils/ingress/wrapper.go:213]: Same ordering comment

File: utils/ingress/ingress.go [hunk 4]
```diff
@@ -11,18 +12,24 @@ import (
...
+	"github.com/argoproj/argo-rollouts/utils/json"
 )
 
 const (
 	// CanaryIngressSuffix is the name suffix all canary ingresses created by the rollouts controller will have
 	CanaryIngressSuffix = "-canary"
 	// ManagedActionsAnnotation holds list of ALB actions that are managed by rollouts
 	ManagedActionsAnnotation = "rollouts.argoproj.io/managed-alb-actions"
+	// ManagedAnnotations holds list of ALB annotations that are managed by rollouts supports multiple annotations
+	ManagedAnnotations = "rollouts.argoproj.io/managed-alb-annotations"
```
Comment (zachaller) [inline utils/ingress/ingress.go:26]: Is this needed with managedRoutes? Don't they container the same info a list of routes that the rollouts controller manages?

**Review by zachaller — CHANGES_REQUESTED** [review]

File: utils/ingress/ingress.go [hunk 5]
```diff
@@ -11,18 +12,24 @@ import (
...
+	"github.com/argoproj/argo-rollouts/utils/defaults"
 	"github.com/argoproj/argo-rollouts/utils/diff"
+	"github.com/argoproj/argo-rollouts/utils/json"
 )
 
 const (
 	// CanaryIngressSuffix is the name suffix all canary ingresses created by the rollouts controller will have
 	CanaryIngressSuffix = "-canary"
 	// ManagedActionsAnnotation holds list of ALB actions that are managed by rollouts
 	ManagedActionsAnnotation = "rollouts.argoproj.io/managed-alb-actions"
```
[the remark this answers is not in the document]
Comment (perenesenko) [reply utils/ingress/ingress.go:24]: Added deprecation info

File: utils/ingress/ingress.go (hunk 1 above)
Comment (zachaller) [reply utils/ingress/ingress.go:174]: Man goland was playing tricks on me apparently

File: utils/ingress/wrapper.go (hunk 2 above)
Comment (perenesenko) [reply utils/ingress/wrapper.go:199]: Thanks! I will fix it.

File: utils/ingress/ingress.go (hunk 4 above)
Comment (perenesenko) [reply utils/ingress/ingress.go:26]: No, It's different functionality here.
There is a code for resetting the albAction for non existing rollouts anymore.

File: utils/ingress/wrapper.go (hunk 3 above)
Comment (perenesenko) [reply utils/ingress/wrapper.go:213]: implemented func SortHttpPaths

**Review by zachaller — APPROVED** [review]

**Review by leoluz — APPROVED** [review]

**Pull request merged**